diff --git a/hopper/setup.py b/hopper/setup.py
index 87f6f45..f51986b 100755
--- a/hopper/setup.py
+++ b/hopper/setup.py
@@ -188,14 +188,22 @@ def _write_ninja_file(path,
     flags = [f'cflags = {" ".join(cflags)}']
     flags.append(f'post_cflags = {" ".join(post_cflags)}')
     if with_cuda:
-        flags.append(f'cuda_cflags = {" ".join(cuda_cflags)}')
-        flags.append(f'cuda_post_cflags = {" ".join(cuda_post_cflags)}')
         cuda_post_cflags_sm80 = [s if s != 'arch=compute_90a,code=sm_90a' else 'arch=compute_80,code=sm_80' for s in cuda_post_cflags]
+        cuda_post_cflags_sm80 += ['-gencode', 'arch=compute_80,code=compute_80']            ## note(zhuobin): add PTX for gwatch instrumentation
         flags.append(f'cuda_post_cflags_sm80 = {" ".join(cuda_post_cflags_sm80)}')
+
         cuda_post_cflags_sm80_sm90 = cuda_post_cflags + ['-gencode', 'arch=compute_80,code=sm_80']
+        cuda_post_cflags_sm80_sm90 += ['-gencode', 'arch=compute_80,code=compute_80']        ## note(zhuobin): add PTX for gwatch instrumentation
+        cuda_post_cflags_sm80_sm90 += ['-gencode', 'arch=compute_90a,code=compute_90a']      ## note(zhuobin): add PTX for gwatch instrumentation
         flags.append(f'cuda_post_cflags_sm80_sm90 = {" ".join(cuda_post_cflags_sm80_sm90)}')
+
         cuda_post_cflags_sm100 = [s if s != 'arch=compute_90a,code=sm_90a' else 'arch=compute_100a,code=sm_100a' for s in cuda_post_cflags]
+        cuda_post_cflags_sm100 += ['-gencode', 'arch=compute_100a,code=compute_100a']       ## note(zhuobin): add PTX for gwatch instrumentation
         flags.append(f'cuda_post_cflags_sm100 = {" ".join(cuda_post_cflags_sm100)}')
+
+        cuda_post_cflags += ['-gencode', 'arch=compute_90a,code=compute_90a']               ## note(zhuobin): add PTX for gwatch instrumentation
+        flags.append(f'cuda_cflags = {" ".join(cuda_cflags)}')
+        flags.append(f'cuda_post_cflags = {" ".join(cuda_post_cflags)}')
     flags.append(f'cuda_dlink_post_cflags = {" ".join(cuda_dlink_post_cflags)}')
     flags.append(f'ldflags = {" ".join(ldflags)}')
 
